### ISSUE TITLE
Pipeline stuck

### ISSUE THREAD

**Issue by bq-zhao** [ENCODE-DCC/atac-seq-pipeline#280]
## **Describe the bug**
Hi team,

My submission gets stuck at the step "2020-09-xx 10:57:34,626|caper.cromwell_workflow_monitor|INFO| Task: id=ee393982-c762-4e1f-aba5-df9c37d17119, task=atac.read_genome_tsv:-1, retry=0, status=WaitingForReturnCode" for hours and won't proceed.

Thanks a lot!
Bingqing

## **OS/Platform**
- OS/Platform: Stanford SCG cluster
- Conda version: 4.8.2
- Pipeline version: 1.7.0
- Caper version: 1.3.3

## **Caper configuration file**

backend=slurm
slurm-account=default

 DO NOT use /tmp here
 Caper stores all important temp files and cached big data files here
 If not defined, Caper will make .caper_tmp/ on your local output directory
 which is defined by out-dir, --out-dir or $CWD
 Use a local absolute path here
tmp-dir=

cromwell=/home/bqzhao/.caper/cromwell_jar/cromwell-47.jar
womtool=/home/bqzhao/.caper/womtool_jar/womtool-47.jar
```

## **Input JSON file**
{
  "atac.title": "RUN16_16_human adipose_no dounce in HB_20um",
  "atac.description": "This is a template input JSON for paired ended sample.",
  "atac.pipeline_type": "atac",
  "atac.align_only": false,
  "atac.true_rep_only": false,
  "atac.genome_tsv": "/labs/mpsnyder/bqzhao/Genomes/hg38/hg38.tsv",
  "atac.paired_end": true,
  "atac.fastqs_rep1_R1": [
    "/labs/mpsnyder/bqzhao/MoTrPAC_BZ/RUN16/200902_A00509_0121_AHNCVJDRXX_L1-MoTrPAC_ATAC_no_dounce_test_pool/stage0_bcl2fastq/MoTrPAC_ATAC_no_dounce_test_pool_16_CGAGGCTG-TCACGTTC_S16_L001_R1_001.fastq.gz",
    "/labs/mpsnyder/bqzhao/MoTrPAC_BZ/RUN16/200902_A00509_0121_AHNCVJDRXX_L2-MoTrPAC_ATAC_no_dounce_test_pool/stage0_bcl2fastq/MoTrPAC_ATAC_no_dounce_test_pool_16_CGAGGCTG-TCACGTTC_S16_L001_R1_001.fastq.gz"
  ],
  "atac.fastqs_rep1_R2": [
    "/labs/mpsnyder/bqzhao/MoTrPAC_BZ/RUN16/200902_A00509_0121_AHNCVJDRXX_L1-MoTrPAC_ATAC_no_dounce_test_pool/stage0_bcl2fastq/MoTrPAC_ATAC_no_dounce_test_pool_16_CGAGGCTG-TCACGTTC_S16_L001_R2_001.fastq.gz",
    "/labs/mpsnyder/bqzhao/MoTrPAC_BZ/RUN16/200902_A00509_0121_AHNCVJDRXX_L2-MoTrPAC_ATAC_no_dounce_test_pool/stage0_bcl2fastq/MoTrPAC_ATAC_no_dounce_test_pool_16_CGAGGCTG-TCACGTTC_S16_L001_R2_001.fastq.gz"
  ],
  "atac.auto_detect_adapter": false,
  "atac.multimapping": 4,
  "atac.align_cpu": 4
}

## **Troubleshooting result**

PASTE TROUBLESHOOTING RESULT HERE
(encode-atac-seq-pipeline) [bqzhao@smsh11dsu-srcf-d15-38 .caper]$ caper debug 17477672
2020-09-23 11:20:32,040|caper.server_heartbeat|ERROR| Failed to read from a heartbeat file. ~/.caper/default_server_heartbeat
Traceback (most recent call last):
  File "/home/bqzhao/miniconda3/envs/encode-atac-seq-pipeline/bin/caper", line 13, in <module>
    main()
  File "/home/bqzhao/miniconda3/envs/encode-atac-seq-pipeline/lib/python3.7/site-packages/caper/cli.py", line 683, in main
    client(parsed_args)
  File "/home/bqzhao/miniconda3/envs/encode-atac-seq-pipeline/lib/python3.7/site-packages/caper/cli.py", line 287, in client
    subcmd_troubleshoot(c, args)
  File "/home/bqzhao/miniconda3/envs/encode-atac-seq-pipeline/lib/python3.7/site-packages/caper/cli.py", line 553, in subcmd_troubleshoot
    cm = get_single_cromwell_metadata_obj(caper_client, args, 'troubleshoot/debug')
  File "/home/bqzhao/miniconda3/envs/encode-atac-seq-pipeline/lib/python3.7/site-packages/caper/cli.py", line 501, in get_single_cromwell_metadata_obj
    wf_ids_or_labels=args.wf_id_or_label, embed_subworkflow=True
  File "/home/bqzhao/miniconda3/envs/encode-atac-seq-pipeline/lib/python3.7/site-packages/caper/caper_client.py", line 132, in metadata
    embed_subworkflow=embed_subworkflow,
  File "/home/bqzhao/miniconda3/envs/encode-atac-seq-pipeline/lib/python3.7/site-packages/caper/cromwell_rest_api.py", line 167, in get_metadata
    workflows = self.find(workflow_ids, labels)
  File "/home/bqzhao/miniconda3/envs/encode-atac-seq-pipeline/lib/python3.7/site-packages/caper/cromwell_rest_api.py", line 243, in find
    r = self.__request_get(CromwellRestAPI.ENDPOINT_WORKFLOWS, params=params)
  File "/home/bqzhao/miniconda3/envs/encode-atac-seq-pipeline/lib/python3.7/site-packages/caper/cromwell_rest_api.py", line 31, in wrapper
    raise requests.exceptions.ConnectionError(message) from None
requests.exceptions.ConnectionError: HTTPConnectionPool(host='localhost', port=8000): Max retries exceeded with url: /api/workflows/v1/query?additionalQueryResultFields=labels&includeSubworkflows=True (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x2b1fcd99a320>: Failed to establish a new connection: [Errno 111] Connection refused'))

Failed to connect to Cromwell server. Check if Caper server is running. Also check if hostname and port are correct. method=GET, url=http://localhost:8000/api/workflows/v1/query?additionalQueryResultFields=labels&includeSubworkflows=True

**Comment by leepc12** [comment]
How did you submit a workflow? Please paste a full command line here. It should be something like 
```
sbatch ... --wrap "caper run ..."
```

**Comment by bq-zhao** [comment]
conda activate encode-atac-seq-pipeline

ATACSRC=~/ATAC_PIPELINE/atac-seq-pipeline
OUTDIR=.
cd ${OUTDIR}
JSON_DIR=run

for json in $(ls ${JSON_DIR}); do

  INPUT_JSON=${JSON_DIR}/${json}
  
  JOB_NAME=$(basename ${INPUT_JSON} | sed "s/\.json.*//")

  sbatch -A mpsnyder -J ${JOB_NAME} --export=ALL --mem 10G -t 4-0 --wrap "caper run ${ATACSRC}/atac.wdl -i ${INPUT_JSON}"
done



Thank you!

**Comment by leepc12** [comment]
Not sure if you've run the above command in a shell script. `conda activate` won't work in a shell script.
Try with `source activate` (deprecating but still working).

Also, pick one INPUT_JSON and run it alone and post `cromwell.out` on your CWD.

**Comment by linzhangTuesday** [comment]
Hi not sure if this is resolved. I also got very similar issue of pipeline stuck.

I was trying the ENCSR356KRQ_subsampled.json by 
`caper run atac.wdl -I ENCSR356KRQ_subsampled.json` on an interactive sge node of h_data=16G, h_vem=64G, -pe shared 4

Then it got stuck as below for ~2 hours before I cancel it. 
"2021-02-XX 14:06:56,585|caper.cromwell_workflow_monitor|INFO| Task: id=4ce23110-eb8f-44ad-9bf4-3d03ce393c61, task=atac.align_mito:1, retry=0, status=WaitingForReturnCode"

The last line in cromwell.out is 
`2021-02-XX 14:06:56,572 cromwell-system-akka.dispatchers.backend-dispatcher-141 INFO  - BackgroundConfigAsyncJobExecutionActor [UUID(4ce23110)atac.align_mito:1:1]: Status change from - to WaitingForReturnCode`

Also I am not sure how to use `sbatch ... --wrap "caper run ...` to submit the job as this is not slurm. 

Any suggestions will be very much appreciated!

**Comment by leepc12** [comment]
@linzhangTuesday: You need to use `echo "caper run ..." | qsub ...` instead of `sbatch ... --wrap "caper run ..."`.
If you run a pipeline on an interactive node then caper should run with a `local` backend (not `sge` backend).
Please post any information helpful for debugging or post a new issue.

**Comment by linzhangTuesday** [comment]
@leepc12  Thanks for the quick response. I have tried `caper init local`, it completed once for the sample json ENCSR356KRQ_subsampled.json. However, when I tried it on my own samples and used `caper init local`. It stucked at a similar place (Please see below the place it stopped). This makes me thinking this is due to something else, which feels a bit random.
`2021-02-XX 08:17:10,239|caper.cromwell_workflow_monitor|INFO| Task: id=ce014d2f-e98a-4a11-a6f6-0b5a36f0843f, task=atac.align_mito:1, retry=0, status=WaitingForReturnCode`

Thanks for the advice on using `echo "caper run ..." | qsub` to submit to sge.
I tried `echo "caper run atac.wdl -i ENCSR356KRQ_subsampled.json" | qsub -V -N ATAC_demo -l h_rt=24:00:00 -l h_vmem=128G`. However, the working directory is empty, while I found ATAC_demo.e6624558 file in my home directory (I left local-loc-dir empty in the ~/.caper/default.conf file, so I suppose it would write in the working directory)

It also reports error of `2021-02-X 12:27:51,774|caper.cromwell|ERROR| RC=127
STDOUT/STDERR=java: symbol lookup error: java: undefined symbol: JLI_InitArgProcessing

Womtool validation failed.`

Another question is that is if I am using `caper init local` when running in an interactive session, this modified the ~/.caper/default.conf file to `backend=local`. Would this influence qsub job to sge at the same time?

Thanks and please let me what additional information and file might help debugging, or if it is preferred to open a new issue.

**Comment by leepc12** [comment]
That looks like a JAVA issue please check if your JAVA is properly installed and check its version.
If not defined, `local-loc-dir` defaults to `local-out-dir + "/.caper_tmp/"`.
Define `local-out-dir` in your configuration, this will be the root of all of your outputs.
`caper init` initializes (deletes and make a new) the configuration file. It does not touch any running job on the cluster.

**Comment by linzhangTuesday** [comment]
Hi @leepc12  Thanks a lot for the quick response. The java version is 1.8.0_112, and I was running in an interactive session on the server so I assume that JAVA was installed correctly. Please see below:

java -version
openjdk version "1.8.0_112"
OpenJDK Runtime Environment (Zulu 8.19.0.1-linux64) (build 1.8.0_112-b16)
OpenJDK 64-Bit Server VM (Zulu 8.19.0.1-linux64) (build 25.112-b16, mixed mode)

It is still stuck at the same place of the pipeline as below. I really don't know what to do.
caper.cromwell_workflow_monitor|INFO| Task: id=ce014d2f-e98a-4a11-a6f6-0b5a36f0843f, task=atac.align_mito:1, retry=0, status=WaitingForReturnCode

I am trying to run for three reps, total 18.1 GB of fastq.gz files. Is it possible that is a memory issue? as I only requested of h_data=16G, h_vem=64G, -pe shared 4 for the interactive session.

**Comment by Kyung-TaeLee** [comment]
Hi @leepc12. I am also having same issue here. I am using CentOS Linux release 7.4.1708 (Core), lustre file system. I installed encode-atac-seq-pipeline using the following commands

『install_conda_env.sh mamba
update_conda_env.sh
pip install caper croo』

When I ran the program, I entered node it self using ssh command (ssh node_name)
and the commandline below

『caper run atac.wdl -i input.json』

and the process is stuck at the same step mentioned above by @linzhangTuesday 

『2021-03-08 17:42:52,350|caper.cromwell_workflow_monitor|INFO| Task: id=f9504021-6aac-45ba-b5e7-3b857f538119, task=atac.read_genome_tsv:-1, retry=0, status=WaitingForReturnCode』

Weird thing is that about 2~3 weeks ago, encode-atac-seq-pipeline worked fine but now the process is stuck at "atac.read_genome_tsv" step. Another weird thing is after a few trial of stopping and re-starting the same process, sometimes it works until the end and generates correct output. I thought that this weird behavior is related with updates on encode-atac-seq-pipeline so I completely removed the conda environment and re-installed it but the same issues happened.
I also tried to activate conda environment using “source” commandline but following error message was printed out

『source activate encode-atac-seq-pipeline
-bash: activate: No such file or directory』

I also tried different backend for caper (local, pbs, slurm). “local” and “slurm” backend is stuck at the same step and for the pbs, following error messages were printed out.

『021-03-09 10:48:49,265|caper.cromwell|INFO| Workflow failed. Auto-troubleshooting...
* Started troubleshooting workflow: id=413717f0-796d-4bb1-89e9-6c32678aab92, status=Failed
* Found failures JSON object.
[
    {
        "message": "Workflow failed",
        "causedBy": [
            {
                "causedBy": [],
                "message": "Job atac.read_genome_tsv:NA:1 exited with return code 241 which has not been declared as a valid return code. See 'continueOnReturnCode' runtime attribute for more details."
            }
        ]
    }
]
* Recursively finding failures in calls (tasks)...

==== NAME=atac.read_genome_tsv, STATUS=Failed, PARENT=
SHARD_IDX=-1, RC=241, JOB_ID=56996
START=2021-03-09T01:48:32.167Z, END=2021-03-09T01:48:33.642Z
STDOUT=/lustre/export/home/kyungtae/dataset/alphaArrestin_splicing/result/2021_02_17_atac_25ul_tn5/atac_tn5_adapter/atac/413717f0-796d-4bb1-89e9-6c32678aab92/call-read_genome_tsv/execution/stdout
STDERR=/lustre/export/home/kyungtae/dataset/alphaArrestin_splicing/result/2021_02_17_atac_25ul_tn5/atac_tn5_adapter/atac/413717f0-796d-4bb1-89e9-6c32678aab92/call-read_genome_tsv/execution/stderr
2021-03-09 10:48:49,266|caper.nb_subproc_thread|ERROR| Cromwell failed. returncode=1
2021-03-09 10:48:49,267|caper.cli|ERROR| Check stdout in /lustre/export/home/kyungtae/dataset/alphaArrestin_splicing/result/2021_02_17_atac_25ul_tn5/atac_tn5_adapter/cromwell.out』

It would be great if you can help me to figure it out. I am currently using the latest version of ENCODE-ATAC-seq pipeline downloaded from release pages for github. Thank you

**Comment by leepc12** [comment]
If you are running on an interactive node you should use `local` backend in the conf (`backend=local`). `caper init local` to overwrite settings on the existing conf (please make a backup if you want) and then define necessary parameters like out, loc directories.

`caper` maximizes parallelization by default so serialize all tasks first `caper run ... --max-concurrent-tasks 1` or set `max-concurrent-tasks=1` in your conf `~/.caper/default.conf`.

**Comment by Kyung-TaeLee** [comment]
HI @leepc12 , Thank you so much for the quick reply. I tried "caper init local" and added the "max-concurrent-tasks=1" to the "default.conf file". Contents of  "default.conf.file" is shown below

'''
backend=local

local-hash-strat=path+modtime

local-loc-dir=
max-concurrent-tasks=1

cromwell=/home/kyungtae/.caper/cromwell_jar/cromwell-52.jar
womtool=/home/kyungtae/.caper/womtool_jar/womtool-52.jar
'''
Then re-run the pipeline but pipeline is stuck at the same step agina (read_gemoe.tsv). I will upload cromwell.out and my input.json file for the run.

** I changed the extension of files to "txt" so that I can upload it here. 
[cromwell.out.txt](https://github.com/ENCODE-DCC/atac-seq-pipeline/files/6112695/cromwell.out.txt)
[control.full.json.txt](https://github.com/ENCODE-DCC/atac-seq-pipeline/files/6112696/control.full.json.txt)

Thank you

**Comment by leepc12** [comment]
Looking at the outputs of `read_genome_tsv` there is no `rc` file. I see `rc.tmp` though. Does `sync` work on your interactive node? example:

```bash
leepc12@nandi:~$ sync
leepc12@nandi:~$
```

Please post the `script` file here.

**Comment by Kyung-TaeLee** [comment]
Hi @leepc12 , thank you so much for the reply. I typed in the sync commandline but it never ends. Our server maintenance guy told me that our file system is lustre file system so that I/O processing could be different from other file system, which could be the reason for this aberrant discontinuance in the pipeline. I will upload the script file below. Thank you!


[script.txt](https://github.com/ENCODE-DCC/atac-seq-pipeline/files/6119918/script.txt)

**Comment by leepc12** [comment]
Please remove this line from your python library.
https://github.com/ENCODE-DCC/caper/blob/master/caper/cromwell_backend.py#L407

First locate the file. Replace `__init__.py` with `cromwell_backend.py`.
```
$ python3
> import caper
> caper.__file__
'/users/leepc12/code/caper/caper/__init__.py'

$ ls -l /users/leepc12/code/caper/caper/cromwell_backend.py
-rw-r--r-- 1 leepc12 users 28634 Mar  1 12:58 /users/leepc12/code/caper/caper/cromwell_backend.py
```
And then delete the above line in `cromwell_backend.py`.

**Comment by leepc12** [comment]
Method 2

I think this method is easier than the above one. Make a backend conf file `your_backend_file_to_override_on_sync.conf
`.
```
backend {
  providers {
    Local {
      config {
        script-epilogue = "sleep 10"
      }
    }
  }
}
```

And run caper with this custom backend. This custom backend will override on the existing backend.
So you don't have to modify the source code.
```
$ caper run ... --backend-file your_backend_file_to_override_on_sync.conf
```

**Comment by Kyung-TaeLee** [comment]
Hi @leepc12 , Thank you so much for your support. I tried both method.
![캡처](https://user-images.githubusercontent.com/67738841/110906445-4d737780-834f-11eb-96e7-29567de869c7.PNG)

I inactivated the line as shown above.  Location of the cromwell.backend file is shown below.
![캡처2](https://user-images.githubusercontent.com/67738841/110906581-75fb7180-834f-11eb-9c13-2df6ed4b195f.PNG)

I also tried the 2nd method but the pipeline is still stuck at the same step (read_geom.tsv, especially generating "rc" file). After I cancel the process, "rc"file appears in the directory.

I will also upload cromwell log files and script file in read_genome directory
[cromwell_log.zip](https://github.com/ENCODE-DCC/atac-seq-pipeline/files/6128558/cromwell_log.zip)

[script.txt](https://github.com/ENCODE-DCC/atac-seq-pipeline/files/6128567/script.txt)


Thank you for your help!

**Comment by leepc12** [comment]
You `script.txt` still has `sync` there. Let's override `sync` with `echo' in your `~/.bashrc`. Add the following line to it and re-login.

```bash
alias sync=echo
```

**Comment by Kyung-TaeLee** [comment]
Hi @leepc12 Thank you again for your follow-ups. I changed my .bashrc but am still stuck at the process. I looked into the script file and found that  the "mv" commandline that move the "rc.tmp" to "rc" file is not being executed. So I executed the mv command myself while running the pipeline and after a few seconds, the process move onto the next step. At the next step which is align_mito, however, the process was stuck again and I found that the same mv commnadline that move "rc_tmp" to "rc" file was written in script file but not being executed. I think I need to find the reason why this mv command line is not being executed, which does not seem be caused by "sync" commandline. Would be great if you have some solutions to this. Thank you very much for your patience and all the comments so far.

**Comment by leepc12** [comment]
Here are some command lines arounds that `mv` stuffs. It looks like `sync` was not correctly aliased by `echo`.
```
(
cd /lustre/export/home/kyungtae/programs/atac-seq-pipeline-1.9.2/test/atac/da2baea1-d4cf-4e15-bedd-0cacb4cfd790/call-read_genome_tsv/execution
sync


)
mv /lustre/export/home/kyungtae/programs/atac-seq-pipeline-1.9.2/test/atac/da2baea1-d4cf-4e15-bedd-0cacb4cfd790/call-read_genome_tsv/execution/rc.tmp /lustre/export/home/kyungtae/programs/atac-seq-pipeline-1.9.2/test/atac/da2baea1-d4cf-4e15-bedd-0cacb4cfd790/call-read_genome_tsv/execution/rc
```

First, please check if the aliased `sync` works on your login node.
```bash
$ alias sync=echo
$ sync hello world
hello world
```

Also, can you try with the following one for `your_backend_file_to_override_on_sync.conf`? The only difference is that `Local` is uncapitalized.
```
backend {
  providers {
    local {
      config {
        script-epilogue = "sleep 10"
      }
    }
  }
}
```
Please check `script.txt` of the stuck task `align`.

**Comment by Kyung-TaeLee** [comment]
Thank you @leepc12 . It's finally working fine. By the way, when I installed conda environment using mamba option, undefined symbol error occurred. The error did not come out when I erased and re-installed it without mamba option. Someone else reported the issue in other post so I think you should know about that. Again, thank you for all your support and sharing such a wonderful analysis pipeline for ATAC-seq. Wish you best luck

**Comment by leepc12** [comment]
Which version of the pipeline did you use? I actually added `tbb` to Conda's `requirements.txt`. Can you try re-installing with the conda env installer in the latest release? Please check if `tbb` is defined in `scrtips/requirements.txt`.

I found `libtbb*`s in environment's lib folder. I installed it with `mamba`.
```
(encode-chip-seq-pipeline) leepc12@nandi:/users/leepc12/code/chip-seq-pipeline2$ ll /users/leepc12/miniconda3/envs/encode-chip-seq-pipeline/lib/libtbb*
lrwxrwxrwx 1 leepc12 users      25 Mar 17 13:39 /users/leepc12/miniconda3/envs/encode-chip-seq-pipeline/lib/libtbbmalloc_proxy.so.2 -> libtbbmalloc_proxy.so.2.1*
-rwxrwxr-x 2 leepc12 users   98984 Mar  3 08:18 /users/leepc12/miniconda3/envs/encode-chip-seq-pipeline/lib/libtbbmalloc_proxy.so.2.1*
lrwxrwxrwx 1 leepc12 users      19 Mar 17 13:39 /users/leepc12/miniconda3/envs/encode-chip-seq-pipeline/lib/libtbbmalloc.so.2 -> libtbbmalloc.so.2.1*
-rwxrwxr-x 2 leepc12 users 1217592 Mar  3 08:18 /users/leepc12/miniconda3/envs/encode-chip-seq-pipeline/lib/libtbbmalloc.so.2.1*
lrwxrwxrwx 1 leepc12 users      14 Mar 17 13:39 /users/leepc12/miniconda3/envs/encode-chip-seq-pipeline/lib/libtbb.so.12 -> libtbb.so.12.1*
-rwxrwxr-x 2 leepc12 users 4648032 Mar  3 08:18 /users/leepc12/miniconda3/envs/encode-chip-seq-pipeline/lib/libtbb.so.12.1*
```

https://github.com/ENCODE-DCC/atac-seq-pipeline/issues/313

**Comment by leepc12** [comment]
Closing this since all conda issues are all fixed >=2.0.0.

**Comment by JaskaranNeuro** [comment]
Hello, 

I appreciate the above comments. But I am facing similar issues, which did not went well using the `your_backend_file_to_override_on_sync.conf`. 

Firstly I was having a similar kind of error as @Kyung-TaeLee`s:
![Screenshot from 2024-07-04 15-56-44](https://github.com/ENCODE-DCC/atac-seq-pipeline/assets/147872755/1b1b7c3c-683b-4dcf-9fa3-25cfc1a6078d)

When I tried the @leepc12s solutions (both methods), and entered the command `caper run atac.wdl -i "ATAC_seq_for_TLR3.json" --conda --max-concurrent-tasks 1 --backend-file your_backend_file_to_override_on_sync.conf`, The following error is now being retrieved:
![Screenshot from 2024-07-04 16-00-27](https://github.com/ENCODE-DCC/atac-seq-pipeline/assets/147872755/b717a60a-e423-4d27-aa9c-51b5bfb18a91)

The cromwell.out file: 
[cromwell.out.txt](https://github.com/user-attachments/files/16102511/cromwell.out.txt)

I`ll really appreciate any advice on that.
